src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs

```diff
@@ -704,83 +672,439 @@ internal enum CustomAttributeEncoding : int
+            {
+                ushort val = GetProlog();
+                return val == 0x0001;
+            }
+
+            public string? GetString()
+            {
+                byte packedLengthBegin = PeekData(sizeof(byte))[0];
+
+                // Check if the embedded string indicates a 'null' string (0xff).
+                if (packedLengthBegin == 0xff) // ECMA 335- II.23.3
+                {
+                    // Consume the indicator.
+                    ReadData(1);
+                    return null;
+                }
+
+                // Not a null string, return a non-null string value.
+                // The embedded string a UTF-8 prefixed by an ECMA-335 packed integer.
+                int length = GetPackedLength(packedLengthBegin);
+                if (length == 0)
+                {
+                    return string.Empty;
+                }
+
+                ReadOnlySpan<byte> utf8ByteSpan = ReadData(length);
+                return Encoding.UTF8.GetString(utf8ByteSpan);
+            }
+
+            private int GetPackedLength(byte firstByte)
+            {
+                if ((firstByte & 0x80) == 0)
+                {
+                    // Consume one byte.
+                    ReadData(1);
+                    return firstByte & 0x7f;
+                }
+
+                int len;
+                ReadOnlySpan<byte> data;
+                if ((firstByte & 0xC0) == 0x80)
+                {
+                    // Consume the bytes.
+                    data = ReadData(2);
+                    len = (data[0] & 0x3f) << 8;
+                    return len + data[1];
+                }
+
+                if ((firstByte & 0xE0) == 0xC0)
+                {
+                    // Consume the bytes.
+                    data = ReadData(4);
+                    len = (data[0] & 0x1f) << 24;
+                    len += data[1] << 16;
+                    len += data[2] << 8;
+                    return len + data[3];
+                }
+
+                throw new OverflowException();
+            }
         }
+    }
 
-        public CustomAttributeEncodedArgument CustomAttributeEncodedArgument => m_encodedArgument;
+    internal sealed class CustomAttributeCtorParameter(CustomAttributeType type)
+    {
+        public CustomAttributeType CustomAttributeType => type;
+        public CustomAttributeEncodedArgument? EncodedArgument { get; set; }
     }
 
-    [StructLayout(LayoutKind.Auto)]
-    internal readonly struct CustomAttributeType
+    internal sealed class CustomAttributeNamedParameter(MemberInfo memberInfo, CustomAttributeEncoding fieldOrProperty, CustomAttributeType type)
     {
+        public MemberInfo MemberInfo => memberInfo;
+        public CustomAttributeType CustomAttributeType => type;
+        public CustomAttributeEncoding FieldOrProperty => fieldOrProperty;
+        public CustomAttributeEncodedArgument? EncodedArgument { get; set; }
+    }
+
+    internal sealed class CustomAttributeType(
```

Comment:
@stephentoub Do we have a preferred naming style for primary constructors? Do we want the instance fields that they create to have `_` prefix?

Reply:
> Do we have a preferred naming style for primary constructors? Do we want the instance fields that they create to have _ prefix?

The same as for any normal constructor.

It's hard to put into guidelines, but I have a preference for us only using primary constructors when either:
a) The body of the class is so small that it's obvious just by looking that a primary ctor parameter is being used, or
b) Having the class declare fields that are initialized to the primary ctor parameter, with the rest of the class then referring to those fields. This lets you make them readonly, for example.